src/storage/src/storage_state.rs

```diff
@@ -4,12 +4,67 @@
 // included in the LICENSE file.
 
 //! Worker-local state for storage timely instances.
+//!
+//! One instance of a [`Worker`], along with its contained [`StorageState`], is
+//! part of an ensemble of storage workers that all run inside the same timely
+//! cluster. We call this worker a _storage worker_ to disambiguate it from
+//! other kinds of workers, potentially other components that might be sharing
+//! the same timely cluster.
+//!
+//! ## Controller and internal communication
+//!
+//! A worker receives _external_ [`StorageCommands`](StorageCommand) from the
+//! storage controller, via a channel. Storage workers also share an _internal_
+//! control/command fabric ([`internal_control`](crate::internal_control)).
+//! Internal commands go through a `Sequencer` dataflow that ensures that all
+//! workers receive all commands in the same consistent order.
+//!
+//! We need to make sure that commands that cause dataflows to be rendered are
+//! processed in the same consistent order across all workers because timely
+//! requires this. To achieve this, we make sure that only internal commands can
+//! cause dataflows to be rendered. External commands (from the controller)
+//! cause internal commands to be broadcast (by only one worker), to get
+//! dataflows rendered.
+//!
+//! The internal command fabric is also used to broadcast messages from a local
+//! operator/worker to all workers. For example, when we need to tear down and
+//! restart a dataflow on all workers when an error is encountered.
+//!
+//! ## Async Storage Worker
+//!
+//! The storage worker has a companion [`AsyncStorageWorker`] that must be used
+//! when running code that requires `async`. This is needed because a timely
+//! main loop cannot run `async` code.
+//!
+//! ## Example flow of commands for `CreateSources`
```

Comment:
I don't like that this is as complicated as it is, but I also don't see immediate better options. Most of the complexity comes from the extra hop through the async worker while the internal/external command split actually seems nice.

We could do more involved things where we replace the fixed-function `Sequencer` with our own dataflow that does the same thing but also has additional operators like resolving resume uppers using async and the like. I opted to not do that for now to keeps things tractable.

Reply:
Could we move the responsibility for polling the command stream to the async worker? (Making run-client responsible for just the timely stuff.)

It would make the command flow at least somewhat more "linear", though I can imagine ownership of the state being a little tricky...

Reply:
> though I can imagine ownership of the state being a little tricky...

It would be decently tricky, because timely (and our usage of it) really is all in on the whole a-worker-is-only-one-thread-and-we-use-Rc-for-sharing thing.